Portforwarding is broken

It fails consistnently in openshift 3.x, with an error like:

    Caused by: java.util.concurrent.RejectedExecutionException: Task io.dekorate.deps.okhttp3.RealCall$AsyncCall@5e20ab3f rejected from java.util.concurrent.ThreadPoolExecutor@4846507f[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 0]

Noticed that when I tried to upgrade decorate to 4.5.1.

Used git bisect and found that the offending commit is: 6c20171e4b4ba21d8aa30fb711490dffe207c682

Need to fix the issue introduced in this commit and add test to prevent this from happening again.

--- a/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java
+++ b/kubernetes-client/src/main/java/io/fabric8/kubernetes/client/KubernetesClient.java
@@ -166,4 +166,5 @@ public interface KubernetesClient extends Client {
 
   SubjectAccessReviewDSL subjectAccessReviewAuth();
 
+
 }

--- a/kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java
+++ b/kubernetes-itests/src/test/java/io/fabric8/openshift/UserImpersonationIT.java
@@ -19,6 +19,7 @@ import io.fabric8.kubernetes.api.model.ServiceAccount;
 import io.fabric8.kubernetes.api.model.ServiceAccountBuilder;
 import io.fabric8.kubernetes.api.model.rbac.*;
 import io.fabric8.kubernetes.client.RequestConfig;
+import io.fabric8.kubernetes.client.RequestConfigBuilder;
 import io.fabric8.openshift.api.model.ProjectRequest;
 import io.fabric8.openshift.api.model.User;
 import io.fabric8.openshift.client.OpenShiftClient;
@@ -107,28 +108,30 @@ public class UserImpersonationIT {
 
   @Test
   public void should_be_able_to_return_service_account_name_when_impersonating_current_user() {
-    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
-    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
-    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
-    requestConfig.setImpersonateExtras(Collections.singletonMap("scopes", Arrays.asList("cn=jane","ou=engineers","dc=example","dc=com")));
+    RequestConfig requestConfig = new RequestConfigBuilder()
+    .withImpersonateUsername(SERVICE_ACCOUNT)
+    .withImpersonateGroups("system:authenticated", "system:authenticated:oauth")
+    .withImpersonateExtras(Collections.singletonMap("scopes", Arrays.asList("cn=jane","ou=engineers","dc=example","dc=com"))).build();
 
-    User user = client.currentUser();
+    User user = client.withRequestConfig(requestConfig).call(c-> c.currentUser());
     assertThat(user.getMetadata().getName()).isEqualTo(SERVICE_ACCOUNT);
   }
 
 
   @Test
   public void should_be_able_to_create_a_project_impersonating_service_account() {
-    RequestConfig requestConfig = client.getConfiguration().getRequestConfig();
-    requestConfig.setImpersonateUsername(SERVICE_ACCOUNT);
-    requestConfig.setImpersonateGroups("system:authenticated", "system:authenticated:oauth");
-    requestConfig.setImpersonateExtras(Collections.singletonMap("scopes", Collections.singletonList("development")));
+    RequestConfig requestConfig = new RequestConfigBuilder()
+    .withImpersonateUsername(SERVICE_ACCOUNT)
+    .withImpersonateGroups("system:authenticated", "system:authenticated:oauth")
+    .withImpersonateExtras(Collections.singletonMap("scopes", Collections.singletonList("development")))
+    .build();
+
     // Create a project
-    ProjectRequest projectRequest = client.projectrequests().createNew()
+    ProjectRequest projectRequest = client.withRequestConfig(requestConfig).call(c -> c.projectrequests().createNew()
       .withNewMetadata()
       .withName(NEW_PROJECT)
       .endMetadata()
-      .done();
+      .done());
 
     // Grab the requester annotation
     String requester = projectRequest.getMetadata().getAnnotations().get("openshift.io/requester");

--- a/openshift-client/src/main/java/io/fabric8/openshift/client/DefaultOpenShiftClient.java
+++ b/openshift-client/src/main/java/io/fabric8/openshift/client/DefaultOpenShiftClient.java
@@ -69,11 +69,12 @@ import okhttp3.OkHttpClient;
 import java.io.InputStream;
 import java.net.MalformedURLException;
 import java.net.URL;
-import java.util.Collection;
-import java.util.List;
-import java.util.Objects;
+import java.util.*;
 
 public class DefaultOpenShiftClient extends BaseClient implements NamespacedOpenShiftClient {
+
+  private static final String API_GROUPS_ENABLED = "API_GROUPS_ENABLED";
+  private static final Map<String, Boolean> API_GROUPS_ENABLED_PER_URL = new HashMap<>();
 
   private URL openShiftUrl;
   private NamespacedKubernetesClient delegate;
@@ -91,9 +92,9 @@ public class DefaultOpenShiftClient extends BaseClient implements NamespacedOpen
   }
 
   public DefaultOpenShiftClient(final OpenShiftConfig config) throws KubernetesClientException {
-    super(config);
+    super(configWithApiGroupsEnabled(clientWithOpenShiftOAuthInterceptor(config), config));
     try {
-      this.httpClient = clientWithOpenShiftOAuthInterceptor(this.httpClient);
+      this.httpClient = clientWithOpenShiftOAuthInterceptor(this.httpClient, config);
       this.delegate = new DefaultKubernetesClient(this.httpClient, config);
       this.openShiftUrl = new URL(config.getOpenShiftUrl());
     } catch (MalformedURLException e) {
@@ -102,14 +103,30 @@ public class DefaultOpenShiftClient extends BaseClient implements NamespacedOpen
   }
 
   public DefaultOpenShiftClient(OkHttpClient httpClient, OpenShiftConfig config) throws KubernetesClientException {
-    super(httpClient, config);
+    super(httpClient, configWithApiGroupsEnabled(httpClient, config));
     try {
-      this.httpClient = clientWithOpenShiftOAuthInterceptor(httpClient);
+      this.httpClient = clientWithOpenShiftOAuthInterceptor(httpClient, getConfiguration());
       this.delegate = new DefaultKubernetesClient(this.httpClient, config);
       this.openShiftUrl = new URL(config.getOpenShiftUrl());
     } catch (MalformedURLException e) {
       throw new KubernetesClientException("Could not create client", e);
     }
+  }
+
+  private static OpenShiftConfig configWithApiGroupsEnabled(OkHttpClient httpClient, OpenShiftConfig config) {
+    String url = config.getMasterUrl();
+    Boolean openshiftApiGroupsEnabled = API_GROUPS_ENABLED_PER_URL.containsKey(url);
+    if (openshiftApiGroupsEnabled) {
+      return config;
+    }
+
+    if (!config.isDisableApiGroupCheck()) {
+      return config.withOpenshiftApiGroupsEnabled(false);
+    }
+
+    Boolean enabled = OpenshiftAdapterSupport.isOpenShiftAPIGroups(httpClient, url);
+    API_GROUPS_ENABLED_PER_URL.put(url, enabled);
+   return config.withOpenshiftApiGroupsEnabled(enabled);
   }
 
   public static DefaultOpenShiftClient fromConfig(String config) {
@@ -120,12 +137,18 @@ public class DefaultOpenShiftClient extends BaseClient implements NamespacedOpen
     return new DefaultOpenShiftClient(Serialization.unmarshal(is, OpenShiftConfig.class));
   }
 
-  private OkHttpClient clientWithOpenShiftOAuthInterceptor(OkHttpClient httpClient) {
-    httpClient = httpClient.newBuilder().authenticator(Authenticator.NONE).build();
-    OkHttpClient.Builder builder = httpClient.newBuilder();
+  private static OkHttpClient clientWithOpenShiftOAuthInterceptor(Config config) {
+    return clientWithOpenShiftOAuthInterceptor(null, config);
+  }
+
+  private static OkHttpClient clientWithOpenShiftOAuthInterceptor(OkHttpClient httpClient, Config config) {
+    OkHttpClient.Builder builder = httpClient != null ?
+      httpClient.newBuilder().authenticator(Authenticator.NONE) :
+      new OkHttpClient.Builder().authenticator(Authenticator.NONE);
+
     builder.interceptors().clear();
-    return builder.addInterceptor(new OpenShiftOAuthInterceptor(httpClient, OpenShiftConfig.wrap(getConfiguration())))
-      .addInterceptor(new ImpersonatorInterceptor(getConfiguration()))
+    return builder.addInterceptor(new OpenShiftOAuthInterceptor(httpClient, OpenShiftConfig.wrap(config)))
+      .addInterceptor(new ImpersonatorInterceptor(config))
       .addInterceptor(new BackwardsCompatibilityInterceptor())
       .build();
   }

--- a/openshift-client/src/main/java/io/fabric8/openshift/client/OpenShiftClient.java
+++ b/openshift-client/src/main/java/io/fabric8/openshift/client/OpenShiftClient.java
@@ -18,6 +18,7 @@ package io.fabric8.openshift.client;
 
 import io.fabric8.kubernetes.api.model.KubernetesList;
 import io.fabric8.kubernetes.client.KubernetesClient;
+import io.fabric8.kubernetes.client.RequestConfig;
 import io.fabric8.kubernetes.client.VersionInfo;
 import io.fabric8.kubernetes.client.dsl.*;
 import io.fabric8.openshift.api.model.*;
@@ -87,6 +88,8 @@ public interface OpenShiftClient extends KubernetesClient {
   SubjectAccessReviewOperation<CreateableSubjectAccessReview, CreateableLocalSubjectAccessReview, CreateableSelfSubjectAccessReview, CreateableSelfSubjectRulesReview> subjectAccessReviews();
 
   MixedOperation<OpenshiftClusterRoleBinding, OpenshiftClusterRoleBindingList, DoneableOpenshiftClusterRoleBinding, Resource<OpenshiftClusterRoleBinding, DoneableOpenshiftClusterRoleBinding>> clusterRoleBindings();
+
+  FunctionCallable<NamespacedOpenShiftClient> withRequestConfig(RequestConfig requestConfig);
 
   /**
    * Returns the current logged in user details similar to the `oc whoami` command.

--- a/openshift-client/src/main/java/io/fabric8/openshift/client/OpenShiftConfig.java
+++ b/openshift-client/src/main/java/io/fabric8/openshift/client/OpenShiftConfig.java
@@ -110,7 +110,7 @@ public class OpenShiftConfig extends Config {
       kubernetesConfig.getImpersonateUsername(),
       kubernetesConfig.getImpersonateGroups(),
       kubernetesConfig.getImpersonateExtras(),
-      true,
+      false,
       false
       );
   }

--- a/openshift-client/src/main/java/io/fabric8/openshift/client/OpenShiftConfig.java
+++ b/openshift-client/src/main/java/io/fabric8/openshift/client/OpenShiftConfig.java
@@ -42,8 +42,9 @@ public class OpenShiftConfig extends Config {
 
   private String oapiVersion = "v1";
   private String openShiftUrl;
-  private boolean disableApiGroupCheck;
   private long buildTimeout = DEFAULT_BUILD_TIMEOUT;
+  private boolean openshiftApiGroupsEnabled;
+  private boolean disableApiGroupCheck; //If group hasn't been explicitly set.
 
   //This is not meant to be used. This constructor is used only by the generated builder.
   OpenShiftConfig() {
@@ -63,11 +64,13 @@ public class OpenShiftConfig extends Config {
   }
 
   @Buildable(builderPackage = "io.fabric8.kubernetes.api.builder", editableEnabled = false, refs = {@BuildableReference(Config.class)})
-  public OpenShiftConfig(String openShiftUrl, String oapiVersion, String masterUrl, String apiVersion, String namespace, Boolean trustCerts, Boolean disableHostnameVerification, String caCertFile, String caCertData, String clientCertFile, String clientCertData, String clientKeyFile, String clientKeyData, String clientKeyAlgo, String clientKeyPassphrase, String username, String password, String oauthToken, int watchReconnectInterval, int watchReconnectLimit, int connectionTimeout, int requestTimeout, long rollingTimeout, long scaleTimeout, int loggingInterval, Integer maxConcurrentRequestsPerHost, String httpProxy, String httpsProxy, String[] noProxy, Map<Integer, String> errorMessages, String userAgent, TlsVersion[] tlsVersions, long buildTimeout, long websocketTimeout, long websocketPingInterval, String proxyUsername, String proxyPassword, String trustStoreFile, String trustStorePassphrase, String keyStoreFile, String keyStorePassphrase, String impersonateUsername, String[] impersonateGroups, Map<String, List<String>> impersonateExtras) {
+  public OpenShiftConfig(String openShiftUrl, String oapiVersion, String masterUrl, String apiVersion, String namespace, Boolean trustCerts, Boolean disableHostnameVerification, String caCertFile, String caCertData, String clientCertFile, String clientCertData, String clientKeyFile, String clientKeyData, String clientKeyAlgo, String clientKeyPassphrase, String username, String password, String oauthToken, int watchReconnectInterval, int watchReconnectLimit, int connectionTimeout, int requestTimeout, long rollingTimeout, long scaleTimeout, int loggingInterval, Integer maxConcurrentRequestsPerHost, String httpProxy, String httpsProxy, String[] noProxy, Map<Integer, String> errorMessages, String userAgent, TlsVersion[] tlsVersions, long buildTimeout, long websocketTimeout, long websocketPingInterval, String proxyUsername, String proxyPassword, String trustStoreFile, String trustStorePassphrase, String keyStoreFile, String keyStorePassphrase, String impersonateUsername, String[] impersonateGroups, Map<String, List<String>> impersonateExtras, boolean openshiftApiGroupsEnabled, boolean disableApiGroupCheck) {
     super(masterUrl, apiVersion, namespace, trustCerts, disableHostnameVerification, caCertFile, caCertData, clientCertFile, clientCertData, clientKeyFile, clientKeyData, clientKeyAlgo, clientKeyPassphrase, username, password, oauthToken, watchReconnectInterval, watchReconnectLimit, connectionTimeout, requestTimeout, rollingTimeout, scaleTimeout, loggingInterval, maxConcurrentRequestsPerHost, httpProxy, httpsProxy, noProxy, errorMessages, userAgent, tlsVersions, websocketTimeout, websocketPingInterval, proxyUsername, proxyPassword, trustStoreFile, trustStorePassphrase, keyStoreFile, keyStorePassphrase, impersonateUsername, impersonateGroups, impersonateExtras);
     this.oapiVersion = oapiVersion;
     this.openShiftUrl = openShiftUrl;
     this.buildTimeout = buildTimeout;
+    this.openshiftApiGroupsEnabled = openshiftApiGroupsEnabled;
+    this.disableApiGroupCheck = openshiftApiGroupsEnabled ? false : disableApiGroupCheck;
 
     if (this.openShiftUrl == null || this.openShiftUrl.isEmpty()) {
       this.openShiftUrl = URLUtils.join(getMasterUrl(), "oapi", this.oapiVersion);
@@ -106,7 +109,9 @@ public class OpenShiftConfig extends Config {
       kubernetesConfig.getKeyStorePassphrase(),
       kubernetesConfig.getImpersonateUsername(),
       kubernetesConfig.getImpersonateGroups(),
-      kubernetesConfig.getImpersonateExtras()
+      kubernetesConfig.getImpersonateExtras(),
+      true,
+      false
       );
   }
 
@@ -114,12 +119,6 @@ public class OpenShiftConfig extends Config {
     return config instanceof OpenShiftConfig ? (OpenShiftConfig) config : new OpenShiftConfig(config);
   }
 
-  public boolean isOpenShiftAPIGroups(OpenShiftClient openShiftClient) {
-    if (isDisableApiGroupCheck()) {
-      return false;
-    }
-    return OpenshiftAdapterSupport.isOpenShiftAPIGroups(openShiftClient);
-  }
 
   private static String getDefaultOapiVersion(Config config) {
     return Utils.getSystemPropertyOrEnvVar(KUBERNETES_OAPI_VERSION_SYSTEM_PROPERTY, config.getApiVersion());
@@ -136,6 +135,10 @@ public class OpenShiftConfig extends Config {
     } else {
       return URLUtils.join(config.getMasterUrl(), "oapi", getDefaultOapiVersion(config));
     }
+  }
+
+  public OpenShiftConfig withOpenshiftApiGroupsEnabled(boolean openshiftApiGroupsEnabled) {
+    return new OpenShiftConfigBuilder(this).withOpenshiftApiGroupsEnabled(openshiftApiGroupsEnabled).build();
   }
 
   private static boolean isRootURL(String url) {
@@ -182,4 +185,11 @@ public class OpenShiftConfig extends Config {
     this.disableApiGroupCheck = disableApiGroupCheck;
   }
 
+  public boolean isOpenshiftApiGroupsEnabled() {
+    return openshiftApiGroupsEnabled;
+  }
+
+  public void setOpenshiftApiGroupsEnabled(boolean openshiftApiGroupsEnabled) {
+    this.openshiftApiGroupsEnabled = openshiftApiGroupsEnabled;
+  }
 }

--- a/openshift-client/src/main/java/io/fabric8/openshift/client/OpenshiftAdapterSupport.java
+++ b/openshift-client/src/main/java/io/fabric8/openshift/client/OpenshiftAdapterSupport.java
@@ -82,6 +82,30 @@ public class OpenshiftAdapterSupport {
       return false;
     }
 
+    /**
+     * Check if OpenShift API Groups are available
+     * @param httpClient   The httpClient.
+     * @param masterUrl    The master url.
+     * @return             True if the new <code>/apis/*.openshift.io/</code> APIs are found in the root paths.
+     */
+    static boolean isOpenShiftAPIGroups(OkHttpClient httpClient, String masterUrl) {
+      try {
+        Request.Builder requestBuilder = new Request.Builder()
+          .get()
+          .url(URLUtils.join(masterUrl, APIS));
+        Response response = httpClient.newCall(requestBuilder.build()).execute();
+        APIGroupList apiGroupList = Serialization.unmarshal(response.body().string(), APIGroupList.class);
+
+        for (APIGroup apiGroup : apiGroupList.getGroups()) {
+          if (apiGroup.getName().endsWith("openshift.io")) {
+            return true;
+          }
+        }
+      } catch(Exception e) {
+        KubernetesClientException.launderThrowable(e);
+      }
+      return false;
+    }
 
     /**
      * Checks if a custom URL for OpenShift has been used.

--- a/openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/OpenShiftOperation.java
+++ b/openshift-client/src/main/java/io/fabric8/openshift/client/dsl/internal/OpenShiftOperation.java
@@ -25,8 +25,6 @@ import io.fabric8.kubernetes.client.dsl.base.HasMetadataOperation;
 import io.fabric8.kubernetes.client.dsl.base.OperationContext;
 import io.fabric8.kubernetes.client.utils.URLUtils;
 import io.fabric8.kubernetes.client.utils.Utils;
-import io.fabric8.openshift.client.DefaultOpenShiftClient;
-import io.fabric8.openshift.client.OpenShiftClient;
 import io.fabric8.openshift.client.OpenShiftConfig;
 import io.fabric8.openshift.client.OpenShiftConfigBuilder;
 
@@ -44,8 +42,7 @@ public class OpenShiftOperation<T extends HasMetadata, L extends KubernetesResou
     OpenShiftConfig config = OpenShiftConfig.wrap(context.getConfig());
     String oapiVersion = config.getOapiVersion();
     if (Utils.isNotNullOrEmpty(context.getApiGroupName())) {
-      try (OpenShiftClient oc = new DefaultOpenShiftClient(context.getClient(), config)) {
-        if (config.isOpenShiftAPIGroups(oc)) {
+        if (config.isOpenshiftApiGroupsEnabled()) {
           String apiGroupUrl = URLUtils.join(config.getMasterUrl(), "apis", context.getApiGroupName(), oapiVersion);
           String apiGroupVersion = URLUtils.join(context.getApiGroupName(), oapiVersion);
           return context.withConfig(new OpenShiftConfigBuilder(config).withOpenShiftUrl(apiGroupUrl).build()).withApiGroupName(context.getApiGroupName()).withApiGroupVersion(apiGroupVersion);
@@ -53,7 +50,6 @@ public class OpenShiftOperation<T extends HasMetadata, L extends KubernetesResou
           String apiGroupUrl = URLUtils.join(config.getMasterUrl(), "oapi", oapiVersion);
           return context.withConfig(new OpenShiftConfigBuilder(config).withOpenShiftUrl(apiGroupUrl).build()).withApiGroupName(context.getApiGroupName()).withApiGroupVersion(oapiVersion);
         }
-      }
     } else {
       String apiGroupUrl = URLUtils.join(config.getMasterUrl(), "oapi", oapiVersion);
       return context.withConfig(new OpenShiftConfigBuilder(config).withOpenShiftUrl(apiGroupUrl).build()).withApiGroupVersion(oapiVersion);

